### PR TITLE
feat(cutover): tidy-up remainder — dry-run, agent id detection, agentChannelId warn, docs (closes cortex#88)

### PR DESCRIPTION
**Pull request by mellanon**
Remainder of cortex#88 — the four items deferred from cortex#95 (the
quick-win subset of the v1 cutover follow-ups). With this PR landed,
cortex#88 closes.

Refs #88. When this merges, tick these checkboxes on the issue:

- [x] Item 2 — `cortex start --dry-run` advertised in docs but not implemented
- [x] Item 3 — `migrate-config` emits agent IDs as `luna`/`luna-2`/`luna-3` regardless of identity
- [x] Item 4 — `migrate-config` repeats the same `agentChannelId` across agents
- [x] Item 8 — v1 cutover docs missing the `leafnodes[].account` requirement for operator-mode

`closes cortex#88` — issue auto-closes on merge.

## What landed

**Item 2 — `cortex start --dry-run` (`src/cortex.ts` + `src/__tests__/cortex.test.ts`)**

New Commander option `--dry-run` on `cortex start`. Validation
logic extracted into an exported `runDryRun(configPath)` helper so
unit tests can invoke it in-process. On success: prints
`cortex config validation OK — N agents, M renderers, NATS=<url>`
to stdout, exit 0. On failure: prints
`cortex config validation FAILED: <schema-error>` to stderr, exit 2.
No adapter / NATS / plist / daemon startup — short-circuit lives
before `mkdirSync(STATE_DIR)` + `checkSingleton()`. Unblocks the
recipe `scripts/postinstall.sh` has been advertising at line 70.

**Item 3 — agent id detection (`src/cli/cortex/commands/migrate-config-lib.ts`)**

New `detectAgentIdFromRoleHints()` helper scans each
`discord[i].roles[]` for the legacy `agent-<name>` role-resolver
hint (the role whose `users[]` carries the bot's own Discord id).
When found, the stripped name becomes the cortex agent id.

Detection ladder:

  1. Role-resolver hint
  2. `deriveAgentId(agent.name)` (pre-#88 behaviour at index 0)
  3. Numeric variant (`luna-2`, `luna-3`) — last resort for index ≥1
     or hint-derived id collisions

Per-adapter warning surfaces the inference for operator audit.

**Item 4 — shared agentChannelId warning (same file)**

New `detectSharedAgentChannelId()` runs after `buildAgents()`. When
2+ converted agents share the same `presence.discord.agentChannelId`,
emits one warning per distinct shared id:

```
WARN: migrate-config: agents [luna,echo,forge] share agentChannelId
1487029848164536361 — set distinct channels in cortex.yaml for
per-agent log routing
```

The field is NOT blanked — an operator may genuinely want shared
log routing. Warning surfaces, operator decides.

**Item 8 — operator-mode + leafnode docs (`docs/plan-cortex-migration.md`)**

New §9 covering the v1 cutover operator-mode recipe end-to-end:

  - §9.1 Why operator mode demands `account:` on leafnode remotes
  - §9.2 The required YAML injection (placeholder A-prefixed pubkey
    + the `nsc list accounts` / `nsc describe account` invocations
    to find your real one)
  - §9.3 Cutover sequence (snapshot → generate → merge → inject →
    restart → verify cortex via `cortex start --dry-run`)
  - §9.4 Rollback path (no cortex restart needed; bot reconnects)

The §9 placement matches the brief — keeps the existing v1 recipe
context co-located rather than spinning up a separate
`sop-operator-mode-nats.md`.

## Test plan

- [x] `bunx tsc --noEmit` — clean across the repo
- [x] `bun test src/cli/cortex/ src/__tests__/cortex.test.ts` — 255 pass / 0 fail
- [x] Full suite: `bun test` — 2448 pass / 1 skip / 0 fail
- [x] Added focused regression tests:
  - `runDryRun — config validator (cortex#88 item 2)` — success / NATS-disabled / schema-failure
  - `convertBotYaml — agent id detection (cortex#88 item 3)` — hint present (single + multi-discord) / no hint → agent.name / no hint multi → numeric / first-hint-wins / empty users[] ignored
  - `convertBotYaml — shared agentChannelId warning (cortex#88 item 4)` — warn fires for 3 sharers (exact message format verbatim) / warn skipped for distinct ids / warn skipped for single agent

## Notes for review

- Item 3 + Item 4 touch `migrate-config-lib.ts` which cortex#95 also touched. The `rewritePaths()` helper + the `mkdirSync` in `migrate-config.ts` are preserved unchanged.
- The dry-run validator goes through `loadConfigWithAgents` (not directly `CortexConfigSchema.parse`) so it handles BOTH legacy `bot.yaml` and cortex-shape `cortex.yaml` inputs — matches what `cortex start` (the non-dry path) does in production.

### CODE REVIEW

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 1]
```diff
@@ -422,9 +521,44 @@ function buildAgents(
...
+    }
+    if (!id) {
+      // Fallback: agent.name (only safe for the first variant — subsequent
+      // variants need a distinct id, hence the numeric suffix at index ≥1).
+      id = i === 0 ? baseId : `${baseId}-${i + 1}`;
+    }
+    // Last-resort: a role-hint produced an id that collides with one
+    // already claimed by an earlier adapter. Fall through to numeric.
+    while (claimedIds.has(id)) {
+      id = `${baseId}-${variantIds.length + 1 + 1}`;
```
Comment (mellanon) [inline src/cli/cortex/commands/migrate-config-lib.ts:553]: **[suggestion/quality]** Collision-avoidance fallback emits an unexpected numeric suffix.

When two adapters carry the same `agent-X` role hint (operator misconfiguration), the second one falls into this `while` loop. With `variantIds.length === 1` already populated, `${baseId}-${1 + 1 + 1}` evaluates to `luna-3` — but the natural numeric ladder for variant index `i=1` is `luna-2`. Operators will see a missing `luna-2` and a present `luna-3` with no explanation.

Fix: use `${baseId}-${i + 1}` (the natural ladder for this variant's index), and increment if that also collides. Or rebase the formula on `claimedIds.size + 1` to walk forward predictably.

Not a blocker — the path only fires when the operator misconfigured two adapters with the identical `agent-X` hint, which today is rare. Would be cleaner with the brief's recommended warning (see separate comment).

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 2]
```diff
@@ -422,9 +521,44 @@ function buildAgents(
...
+      }
+    }
+    if (!id) {
+      // Fallback: agent.name (only safe for the first variant — subsequent
+      // variants need a distinct id, hence the numeric suffix at index ≥1).
+      id = i === 0 ? baseId : `${baseId}-${i + 1}`;
+    }
+    // Last-resort: a role-hint produced an id that collides with one
+    // already claimed by an earlier adapter. Fall through to numeric.
+    while (claimedIds.has(id)) {
```
Comment (mellanon) [inline src/cli/cortex/commands/migrate-config-lib.ts:552]: **[suggestion/quality]** No warning when two adapters share the same `agent-<X>` hint.

The brief flagged this edge case explicitly: "what if TWO discord adapters have the same `agent-X` hint (operator misconfigured)? Should be detected + warned." Today the collision-avoidance silently falls through to numeric suffix — operator gets `echo, luna-3` (cf. separate comment about the formula) with no signal that they had two `agent-echo` hints.

Suggested fix: when `claimedIds.has(id)` in this `while` loop, push a warning along the lines of `discord[${i}] role hint resolved to "echo" but that id was already claimed by an earlier adapter — falling back to "${variantId}"`. Operator can then audit the duplicate hint in bot.yaml.

There's no test covering this case either — adding one would lock in the warning + the numeric fallback behaviour.

File: docs/plan-cortex-migration.md [hunk 3]
```diff
@@ -649,4 +649,87 @@ The migration is complete when all of:
...
+     should return without auth errors.
+7. **Verify cortex.** `cortex start --config ~/.config/cortex/cortex.yaml --dry-run` confirms the config still parses (cortex#88 item 2); `cortex status` after `launchctl load …cortex.bot.plist` shows the bot connected to the operator-mode server.
+
+### 9.4 Rollback
+
+If step 6 surfaces an unrecoverable error: restore `local.conf` from the
+pre-cutover snapshot (`cp ~/.nats/local.conf.pre-cutover ~/.nats/local.conf`),
+unload/load the launchd plist again, and resume on the mem-resolver
+configuration while debugging the operator-mode setup offline. Cortex
+itself does not need to be restarted — the bot reconnects to the
```
Comment (mellanon) [inline docs/plan-cortex-migration.md:730]: **[suggestion/docs]** §9 doesn't cross-reference cortex#86 (creds-auth) or cortex#98 (trust mesh).

The brief asked for these explicit links so operators reading the operator-mode cutover doc know where the surrounding plumbing lives. Both are directly relevant: §9 talks about leafnode account-binding which is creds-auth's territory (cortex#86), and the trust-mesh issue (cortex#98) is the natural next step after operator-mode is up.

Suggested: add a one-line "See also" footer to §9 or sprinkle inline links. E.g.:

> See cortex#86 for the creds-auth surface that produces these `A...` pubkeys, and cortex#98 for the trust mesh that runs on top of operator-mode.

Not a blocker.

File: src/cortex.ts [hunk 4]
```diff
@@ -913,6 +913,53 @@ function checkSingleton(): void {
...
+export function runDryRun(configPath: string): DryRunResult {
+  const stdout: string[] = [];
+  const stderr: string[] = [];
+  try {
+    const { config, inlineAgents } = loadConfigWithAgents(configPath);
+    // For cortex-shape input `inlineAgents.length` is the canonical agent
+    // count (operator perspective). Legacy bot.yaml carries one singular
+    // `agent:` field — fall back to 1 in that case so the operator sees a
+    // truthful number rather than `0 agents` on a working legacy config.
+    const agentCount = inlineAgents.length > 0 ? inlineAgents.length : 1;
```
Comment (mellanon) [inline src/cortex.ts:945]: **[nit/quality]** Fallback `1` is a friendly fiction, but masks a degenerate case.

`inlineAgents.length > 0 ? inlineAgents.length : 1` — when a legacy `bot.yaml` parses cleanly but `loadConfigWithAgents` returns zero inline agents (degenerate file with no `agent:` key but valid otherwise), the dry-run prints `1 agents` which isn't true. The schema would normally catch this earlier so it's edge, but a more honest fallback is `inlineAgents.length` directly (yielding `0 agents`), or skip the count for legacy and print `agents=(legacy bot.yaml)` style. Not a blocker.

**Comment by mellanon** [comment]
## Full Review Summary: cortex#104

Reviewed by Echo via the CodeReview skill (FullReview workflow). PR closes cortex#88 — the four deferred quick-wins from the v1 cutover (items 2, 3, 4, 8).

### Lens Results

| Lens | Critical | Major | Suggestion | Nit | Verdict |
|------|----------|-------|------------|-----|---------|
| CodeQuality | 0 | 0 | 2 | 1 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | n/a (no API surface touched) |
| Architecture | 0 | 0 | 0 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 0 | 0 | pass |
| Docs (item 8) | 0 | 0 | 1 | 0 | pass |

### Per-item verdict

| Item | Brief check | Verdict |
|------|-------------|---------|
| 2 — `cortex start --dry-run` | `runDryRun` exported, short-circuits before `mkdirSync`/`checkSingleton`/adapter start. Tests cover success / NATS-disabled / schema-failure with exit code AND operator-readable error text. postinstall.sh `--dry-run` reference now real. | pass (1 nit) |
| 3 — `detectAgentIdFromRoleHints` | Logic matches spec: scan `agent-*` roles with non-empty `users[]`, strip prefix, normalize. Tests cover hint / fallback / numeric / first-wins / empty users[] / multi-discord. Brief's flagged edge case — TWO adapters sharing one `agent-X` hint — falls through silently to numeric with an unexpected formula. | pass (2 suggestions) |
| 4 — `detectSharedAgentChannelId` | Warn fires once per distinct shared id, field NOT blanked, exact message text matches issue spec. Tests cover shared / distinct / single-agent. | pass (0 findings) |
| 8 — `plan-cortex-migration.md §9` | §9.1-9.4 land the operator-mode + leafnode-binding recipe with concrete YAML + cutover sequence + rollback. Missing cross-refs to cortex#86 (creds-auth) and cortex#98 (trust mesh) per brief. | pass (1 suggestion) |

### Cross-checks

- **cortex#95 preservation**: `rewritePaths()` at `migrate-config-lib.ts:715` + `mkdirSync(dirname(outPath))` at `migrate-config.ts:156` both intact — no accidental refactor.
- **Conventional commits**: All 4 commits use `feat(scope): ...` or `docs(plan): ...` + `Refs cortex#88 (item N of 4 remaining)`. PR body has `closes cortex#88`.
- **Test suite**: 71/71 pass on the migrate-config tests in-worktree. Engineer reports 2448 / 1 skip / 0 fail repo-wide; 1 skip is the cortex#89 dashboard SPA self-skip.
- **`tsc --noEmit`**: clean (engineer-reported, not re-run in this review).

### Key findings

1. **Collision-fallback formula bug** (`migrate-config-lib.ts:553`): `${baseId}-${variantIds.length + 1 + 1}` produces `luna-3` for the second collision at variant index 1 instead of `luna-2`. Surprises operators who get a non-contiguous numeric ladder. Suggestion only — the path only fires on misconfigured-duplicate-hint, but is also the case for finding #2 below.
2. **No warning on duplicate hints** (`migrate-config-lib.ts:552`): the brief explicitly flagged this edge case — two adapters with identical `agent-X` hint should be detected + warned. Today it falls through silently to numeric.
3. **§9 cross-refs missing** (`plan-cortex-migration.md`): brief asked for explicit links to cortex#86 (creds-auth — produces the `A...` pubkeys) and cortex#98 (trust mesh — sits on top of operator-mode).
4. **Nit on dry-run agent count fallback** (`cortex.ts:945`): legacy `bot.yaml` w/ zero inline agents falls back to printing `1 agents` — degenerate case, but more honest to print actual count.

All findings are suggestion/nit severity. None block merge.

### Overall Verdict: merge

The four items match the spec, tests are thorough, the docs §9 is the right shape and content. The collision case in item 3 (findings 1+2) deserves a follow-up but the production path (each adapter has its own distinct `agent-*` hint) is correctly covered. Recommend merge + a small follow-up ticket for the duplicate-hint warning + formula tidy. Item 8 cross-refs can be added as a doc-touchup in the same follow-up.

— Echo

**Review by mellanon — COMMENTED** [review]
Full review (6 lenses + duplication) — 4 items cleanly match spec; 0 blockers, 0 majors, 3 suggestions + 1 nit. Item 3's collision-fallback (duplicate `agent-X` hints) emits an unexpected numeric ladder + no warning — see inline. Item 8 doc §9 missing cross-refs to cortex#86 / cortex#98 per brief.

verdict: blockers=0 majors=0 suggestions=3 nits=1 — recommend: merge
